### PR TITLE
dub.project: Improve variable expansion logic

### PR DESCRIPTION
**Pull request by CyberShadow**
See https://github.com/dlang/dmd/pull/9275#issuecomment-455861862

### CODE REVIEW

File: source/dub/project.d [hunk 1]
```diff
@@ -1210,6 +1205,91 @@ private string processVars(Project, Package)(string var, in Project project, in
...
+	assertThrown(expandVars!expandVar("$"));
+	assertThrown(expandVars!expandVar("${}"));
+	assertThrown(expandVars!expandVar("$|"));
+	assertThrown(expandVars!expandVar("x$"));
+	assertThrown(expandVars!expandVar("$X"));
+	assertThrown(expandVars!expandVar("${"));
+	assertThrown(expandVars!expandVar("${X"));
+
+	// https://github.com/dlang/dmd/pull/9275
+	assert(expandVars!expandVar("$${DUB_EXE:-dub}") == "${DUB_EXE:-dub}");
```
Comment (jacob-carlborg) [inline source/dub/project.d]: Should we test for other Bash variable syntaxes: https://www.tldp.org/LDP/abs/html/parameter-substitution.html ?

Comment (CyberShadow) [reply source/dub/project.d]: Well, the point of this test is to explicitly check that Dub doesn't do anything crazy here, and it ignores anything following `$$` and passes it through verbatim, whatever the syntax might be. The previous implementation expected very specific syntax to be used even when escaping Dub processing, which was presumptuous and led to wrong behavior. This test (and any other tests we might add in the same vein) are supplemental / illustrative to the behavior tests above, and don't actually add to the test coverage. An implementation of `expandVars` which somehow passes all of these tests but would behave incorrectly with other kinds of Bash variable substitution syntax would be very weird indeed.

Comment (jacob-carlborg) [reply source/dub/project.d]: Ok, good enough for me.

**Comment by ZombineDev** [comment]
> Instead of looking for $ (minding escapes), it uses a regular expression to match what it thinks shell variable use looks like.

@CyberShadow Yes, it is working as intended - it's just that we have a different definition of what's intended. I asked for input on this matter on the newsgroup, but apparently no one cared enough to respond at the time, so we went with the current behavior.

* https://forum.dlang.org/post/exerdovrnivrsloetxtg@forum.dlang.org
* https://github.com/dlang/dub/pull/1392/files#diff-fc1737ce75bc9102f5d5b54be258a268R1188
* https://github.com/dlang/dub/pull/1552

**Comment by CyberShadow** [comment]
@ZombineDev Sorry, what do you mean? Of the unit tests I added, how would you consider an implementation that failed them as "working as intended"?

**Comment by wilzbach** [comment]
I'm going to merge this now as it still passes all the tests of the referenced https://github.com/dlang/dub/pull/1392 which is in master now.

**Pull request merged**